src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+        match self {
+            PcntCountMode::Hold => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_HOLD,
+            PcntCountMode::Increment => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE,
+            PcntCountMode::Decrement => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE,
+        }
+    }
+}
+
+#[doc = " @brief PCNT channel action on control level"]
+#[derive(Debug, Copy, Clone)]
+pub enum PcntControlMode {
+    #[doc = "< Keep current count mode"]
+    Keep,
+    #[doc = "< Invert current count mode (increase -> decrease, decrease -> increase)"]
+    Reverse,
+    #[doc = "< Hold current count value"]
+    Disable,
+}
+
+impl Into<esp_idf_sys::pcnt_ctrl_mode_t> for PcntControlMode {
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        #[doc = "< PCNT watch point event: threshold1 value event"]
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        #[doc = "< PCNT watch point event: threshold0 value event"]
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        #[doc = "< PCNT watch point event: Minimum counter value"]
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        #[doc = "< PCNT watch point event: Maximum counter value"]
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        #[doc = "< PCNT watch point event: counter value zero event"]
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+pub struct PcntConfig<'d> {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<&'d AnyInputPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<&'d AnyInputPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
```

Comment:
As per earlier comment, this is also a peripheral and therefore does not belong here.

Reply:
I'm not sure that channel should be a peripheral. Unlike LEDC timers they can not be shared across pcnt units and each unit has two channels that both operate against the units counter (and limits).

Reply:
It's interesting that v4 implies that the limits could be set differently in each of the channels but in v5 the limits are configured on the unit not the channels.  Maybe I'll try to set up a test using v4 and see what happens if I configure seperate limits on the channels.

Reply:
yea - in v4 if the limits are different between two channels then the limits used are from the last channel configured. (one set of limits per unit)